kasa/smartdevice.py

```diff
@@ -294,13 +297,19 @@ async def update(self):
 
         Needed for methods that are decorated with `requires_update`.
         """
-        req = {}
-        req.update(self._create_request("system", "get_sysinfo"))
+        update_response = {}
+        update_response.update(
+            await self.protocol.query(
+                self.host, self._create_request("system", "get_sysinfo")
+            )
+        )
+
+        if SmartDevice._check_emeter_feature_presence(update_response):
+            update_response.update(
+                await self.protocol.query(self.host, self._create_emeter_request())
+            )
```

Comment:
This would create unnecessary I/O requests on devices that support the emeter. The logic should be:
1. If `_sysinfo` is None (i.e., the device hasn't ever been updated and we don't know if it supports emeter), do fetch only the sysinfo. Check if the `_has_emeter` is now True, and request the emeter information if so.
2. Otherwise (on an already initialized device), if `_has_emeter` is True, the emeter request should be included in the query like it is at the moment.

This way only the initial update will cause two separate I/O requests.

Reply:
So, I had a chance to play around a little more this weekend and it turns out that the fix is actually much more simple than I had originally thought.

On my HS220 device, I was able to determine that:

1) The emeter check returns as expected when it's the only query (which is what led me to the original implementation in this PR)
2) the emeter check returns as expected when combined with other non-system queries (i.e. time, schedule, etc...)

and finally...
3) The emeter check returns as expected when combined with the system query **as long as the emeter query comes first**

Since the handful of other devices that I have (HS210, HS200, KP400) were all working with the original code, I inspected the software version to figure out what the difference is. The HS220's look to be running an older version of the firmware than all the other devices, so I suspect there's probably a bug in older versions of the firmware (1.0.5 seems to be the cutoff) when querying for both emeter and system in the same request.

Reply:
Ah, thanks for the comprehensive analysis! I just missed this when doing a review on the currently changed code pieces.

A question though, what is the expected response for the emeter when it's the only query? My assumption is that HS220 does not support emeter at all => there is no need to do any tricks to add emeter queries for devices without emeter (as reported by the sysinfo)?